src/psyclone/psyir/nodes/acc_directives.py

```diff
@@ -737,6 +689,65 @@ def end_string(self):
         '''
         return "acc end data"
 
+    def tree_update(self):
+        '''
+        Called whenever there is a change in the PSyIR tree below this node.
+        This may mean that the various copy[in/out] clauses need updating if
+        the contents of the data region have changed.
+
+        '''
+        # Ensure that this method does not get called recursively.
+        if self._disable_tree_update:
+            return
+        self._disable_tree_update = True
+
+        # Remove the clauses that we will update.
+        for child in self.children[:]:
+            if isinstance(child,
+                          (ACCCopyInClause, ACCCopyOutClause, ACCCopyClause)):
+                self.children.remove(child)
+
+        # Identify the inputs and outputs to the region (variables that
+        # are read and written).
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.tools import DependencyTools
+        dtools = DependencyTools()
+        in_outs = dtools.get_in_out_parameters(self.children)
+        var_accesses = dtools.variable_access_info
+        readers = in_outs.signatures_read
+        writers = in_outs.signatures_written
+        readwrites_list = in_outs.signatures_readwrite
+        readwrites = set(readwrites_list)
+        readers_list = sorted(list(set(readers) - readwrites))
+        writers_list = sorted(list(set(writers) - readwrites))
+
+        # We now need to create PSyIR references for all of the signatures
+        # and add them as children of the appropriate clauses.
+        nodes_dict = OrderedDict()
+        for sig in readers_list:
+            sig.create_deep_copy_refs(var_accesses[sig].all_accesses[0].node,
```

Comment:
This is a continuation of a previous discussion regarding the "create_deep_copy_refs"

I think that logic belongs in this class (or it parent Directive), but probably also some of the things that are currently inside this tree_udpate should be part of it.

Instead of doing the depanalysis/reference_accesses and iteration here an alternative is to bring those also inside the method, I think it makes a more clear API

`Directive._create_data_movement_deep_copy_refs(): (OrderedDict, OrderedDict, OrderedDict)`
or
`Directive._create_data_movement_deep_copy_refs("in"/"out/"inout"): OrderedDict`

This prevents having to pass and typecheck the OrderDict, and also avoids needing to typecheck the node argument which is currently unchecked and it probably should as it is currently a public method.

Also an intermediate method that calls the `create_data_movement_deep_copy_refs` and puts the returned references on the right clause could be introduced, e.g.: `Directive.create/insert_data_movement_clauses()`
The reason for this is that due to how we implemented clauses, I expect this tree_update method to grow, as thing like reduction clauses, Aidan's tasks, possible atomics would all depend on the directive_body and probably need to be updated by this function, so it would be good to have a sub-method for each to make it clear and overview of what it does.

Reply:
Done.